llvm/include/llvm/CodeGen/GlobalISel/GIMatchTableExecutor.h

```diff
@@ -713,6 +713,27 @@ class GIMatchTableExecutor {
     memcpy(&Ret, MatchTable, sizeof(Ret));
     return Ret;
   }
+
+public:
+  // Faster ULEB128 decoder tailored for the Match Table Executor.
+  //
+  // - Arguments are fixed to avoid mid-function checks.
+  // - Unchecked execution, assumes no error.
+  // - Fast common case handling (1 byte values).
+  LLVM_ATTRIBUTE_ALWAYS_INLINE static uint64_t
+  fastDecodeULEB128(const uint8_t *__restrict MatchTable,
```

Comment:
Do we need to guard the non-standard __restrict in a macro? I don't see any prior art for this

Reply:
Is it enough to indicate the underlying table is actually const?

Reply:
__restrict seems to allow for optimizing away a whole branch. I assume it doesn't reload MatchTable as often because it knows `CurrentIdx` cannot alias it

It probably needs a compiler macro though, I'll look into it

Reply:
If it makes an observable ISA change I think it's worth trying to keep

Reply:
It's very small: https://godbolt.org/z/f9341f1Ea
Though I sometimes wonder why we don't use restrict in more places. Seems like a useful thing in some cases.

I'd say let's do without it for now and start a discussion on Discourse? Using restrict would set a precedent and I think it's better to get community approval before using that in non-runtime code

Reply:
This isn't really worthy of an RFC. We just need another macro in Support/Compiler.h